testng-core/src/main/java/org/testng/internal/ConfigurationGroupMethods.java

```diff
@@ -121,9 +73,37 @@ public List<ITestNGMethod> getBeforeGroupMethodsForGroup(String[] groups) {
     }
   }
 
-  public List<ITestNGMethod> getAfterGroupMethodsForGroup(String group) {
+  public List<ITestNGMethod> getAfterGroupMethods(ITestNGMethod testMethod) {
+    if (testMethod.hasMoreInvocation() || testMethod.getGroups().length == 0) {
+      return Collections.emptyList();
+    }
+
+    Set<String> methodGroups = new HashSet<>(Arrays.asList(testMethod.getGroups()));
+
     synchronized (afterGroupsThatHaveAlreadyRun) {
-      return retrieve(afterGroupsThatHaveAlreadyRun, m_afterGroupsMethods, group);
+      if (m_afterGroupsMap == null) {
+        m_afterGroupsMap = initializeAfterGroupsMap();
+      }
+
+      return methodGroups.stream()
+          .filter(t -> isLastMethodForGroup(t, testMethod))
+          .map(t -> retrieve(afterGroupsThatHaveAlreadyRun, m_afterGroupsMethods, t))
+          .filter(Objects::nonNull)
+          .flatMap(Collection::stream)
+          .filter(
+              afterGroupMethod -> {
+                String[] afterGroupMethodGroups = afterGroupMethod.getAfterGroups();
```

Comment:
For better readability, can we please extract out this lambda as a separate method which can then be invoked from within here?